patch/netlink.patch

```diff
@@ -0,0 +1,13 @@
+--- netlink.c	2021-01-09 19:18:42.109012366 +0100
++++ netlink.c	2021-01-09 16:08:05.098280000 +0100
+@@ -62,8 +62,8 @@
+ 				      nla_data(attrs[WGDEVICE_A_IFNAME]));
+ 	if (!dev)
+ 		return ERR_PTR(-ENODEV);
+-	if (!dev->rtnl_link_ops || !dev->rtnl_link_ops->kind ||
+-	    strcmp(dev->rtnl_link_ops->kind, KBUILD_MODNAME)) {
++	if (!dev->dev.type || !dev->dev.type->name ||
++			strcmp(dev->dev.type->name, KBUILD_MODNAME)) {
```

Comment:
When making this patch, I relied on wireguard-linux-compat source code and kernel 4.4.180 source code.

In the simplest terms:

1. WireGuard [register rtnl_link_ops](https://github.com/WireGuard/wireguard-linux-compat/blob/d6d5caa4f16f0fb9dcadca58e6adf5160f13acbd/src/device.c#L447) with rtnetlink.
2. The function [rtnl_newlink](https://elixir.bootlin.com/linux/v4.4.180/source/net/core/rtnetlink.c#L2180) from rtnetlink is called.
3. The [rtnl_create_link](https://elixir.bootlin.com/linux/v4.4.180/source/net/core/rtnetlink.c#L2180) function is called, which sets the value of [rtnl_link_ops](https://elixir.bootlin.com/linux/v4.4.180/source/net/core/rtnetlink.c#L2132).

On DSM 7.0 (Geminilake in my case) `rtnl_link_ops` is not set, which [causes an error](https://github.com/runfalk/synology-wireguard/issues/66#issuecomment-755479042). However, the `wg_setup` and `wg_newlink` functions from the [link_ops structure](https://github.com/WireGuard/wireguard-linux-compat/blob/c3c51d07959a55602bb3c40bf9b2fb694bcc8f5a/src/device.c#L404) are executed correctly. So I had to find another way to check if the interface is a WireGuard interface.
The `wg_setup` function calls [SET_NETDEV_DEVTYPE](https://elixir.bootlin.com/linux/v4.7/source/include/linux/netdevice.h#L2015), which sets as `type` a [device_type structure](https://github.com/WireGuard/wireguard-linux-compat/blob/c3c51d07959a55602bb3c40bf9b2fb694bcc8f5a/src/device.c#L261) with a `.name` field of constant `KBUILD_MODNAME`. Therefore, in my opinion `dev->dev.type->name` is a good replacement for `dev->rtnl_link_ops->kind` for the interface check. Of course I could be wrong, but it would have to be verified by someone with more experience.